Lib/exception_group.py

```diff
@@ -0,0 +1,197 @@
+                        self.tb_next_map[id(k)] = e_.__traceback__
+            else:
+                self.tb_next_map[id(e)] = e.__traceback__
+
+
+class ExceptionGroup(BaseException):
+    def __init__(self, message, *excs):
+        """ Construct a new ExceptionGroup
+
+        excs: sequence of exceptions
+        tb [optional]: the __traceback__ of this exception group.
+        Typically set when this ExceptionGroup is derived from another.
+        """
+        assert message is None or isinstance(message, str)
+        assert all(isinstance(e, BaseException) for e in excs)
+
+        self.message = message
+        self.excs = excs
+        super().__init__(self.message)
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is set when the exception group is created.
+        # __traceback_group__ and __traceback__ combine to give the full path.
+        self.__traceback_group__ = TracebackGroup(self.excs)
+
+    def project(self, condition, with_complement=False):
+        """ Split an ExceptionGroup based on an exception predicate
+
+        returns a new ExceptionGroup, match, of the exceptions of self
+        for which condition returns True. If with_complement is True,
+        returns another ExceptionGroup for the exception for which
+        condition returns False.
+        match and rest have the same nested structure as self, but empty
+        sub-exceptions are not included. They have the same message,
+        __traceback__, __cause__ and __context__ fields as self.
+
+        condition: BaseException --> Boolean
+        with_complement: Bool  If True, construct also an EG of the non-matches
+        """
+        match = []
+        rest = [] if with_complement else None
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):  # recurse
+                e_match, e_rest = e.project(
+                    condition, with_complement=with_complement)
+                if not e_match.is_empty():
+                    match.append(e_match)
+                if with_complement and not e_rest.is_empty():
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                elif with_complement:
+                    rest.append(e)
+
+        match_exc = ExceptionGroup(
+            self.message, *match).with_traceback(self.__traceback__)
+
+        def copy_metadata(src, target):
+            target.__context__ = src.__context__
+            target.__cause__ = src.__cause__
+
+        copy_metadata(self, match_exc)
+        if with_complement:
+            rest_exc = ExceptionGroup(self.message, *rest).with_traceback(self.__traceback__)
+            copy_metadata(self, rest_exc)
+        else:
+            rest_exc = None
+        return match_exc, rest_exc
+
+    def split(self, type):
+        """ Split an ExceptionGroup to extract exceptions of type E
+
+        type: An exception type
+        """
```

Comment:
Due to the way that project() flattens the exception group, split doesn't match anything if the type is ExceptionGroup:

>>> e1 = ValueError(1)
>>> e2 = ValueError(2)
>>> eg1 = exception_group.ExceptionGroup('eg1', e1, e2)
>>> eg2 = exception_group.ExceptionGroup('eg2', eg1, e1, e2)
>>> match, rest = eg2.split(exception_group.ExceptionGroup)
>>> match
ExceptionGroup(eg2, ())      # <-- empty ExceptionGroup
>>> rest
ExceptionGroup(eg2, (ExceptionGroup(eg1, (ValueError(1), ValueError(2))), ValueError(1), ValueError(2)))

If someone does 

try:
   ....
except ExceptionGroup as e:
    # and e is a nested ExceptionGroup.

then e is just the whole ExceptionGroup and we don't attempt to split further, right? 
So should split() special-case type == ExceptionGroup?

Reply:
Will we have subclasses of ExceptionGroup? If we do, then we could have an exception group with a nested SubExceptionGroup(), and then we split on SubExceptionGroup. Then the solution needs to be inside project() because the match is not necessarily at the top level ExceptionGroup.